node/node_handler.go

```diff
@@ -812,3 +823,64 @@ func (node *Node) ConsensusMessageHandler(msgPayload []byte) {
 	}
 	return
 }
+
+// ExplorerMessageHandler passes received message in node_handler to explorer service
+func (node *Node) ExplorerMessageHandler(payload []byte) {
+	if len(payload) == 0 {
+		return
+	}
+	msg := &msg_pb.Message{}
+	err := protobuf.Unmarshal(payload, msg)
+	if err != nil {
+		utils.GetLogger().Error("Failed to unmarshal message payload.", "err", err)
+		return
+	}
+
+	if msg.Type == msg_pb.MessageType_COMMITTED {
+
+		recvMsg, err := consensus.ParsePbftMessage(msg)
+		if err != nil {
+			utils.GetLogInstance().Debug("[Explorer] onCommitted unable to parse msg", "error", err)
+			return
+		}
+
+		aggSig, mask, err := node.Consensus.ReadSignatureBitmapPayload(recvMsg.Payload, 0)
+		if err != nil {
+			utils.GetLogInstance().Debug("[Explorer] readSignatureBitmapPayload failed", "error", err)
+			return
+		}
+
+		// check has 2f+1 signatures
+		if count := utils.CountOneBits(mask.Bitmap); count < node.Consensus.Quorum() {
+			utils.GetLogInstance().Debug("[Explorer] not have enough signature", "need", node.Consensus.Quorum(), "have", count)
+			return
+		}
+
+		blockNumHash := make([]byte, 8)
+		binary.LittleEndian.PutUint64(blockNumHash, recvMsg.BlockNum)
+		commitPayload := append(blockNumHash, recvMsg.BlockHash[:]...)
+		if !aggSig.VerifyHash(mask.AggregatePublic, commitPayload) {
+			utils.GetLogInstance().Debug("[Explorer] Failed to verify the multi signature for commit phase", "msgBlock", recvMsg.BlockNum)
+			return
+		}
+		block := node.Consensus.PbftLog.GetBlockByHash(recvMsg.BlockHash)
+
+		// Dump new block into level db.
+		utils.GetLogInstance().Info("[Explorer] Committing block into explorer DB", "msgBlock", recvMsg.BlockNum)
+		explorer.GetStorageInstance(node.SelfPeer.IP, node.SelfPeer.Port, true).Dump(block, block.NumberU64())
+	} else if msg.Type == msg_pb.MessageType_ANNOUNCE {
+
+		recvMsg, err := consensus.ParsePbftMessage(msg)
+		if err != nil {
+			utils.GetLogInstance().Debug("[Explorer] onAnnounce unable to parse msg", "error", err)
+			return
+		}
+		block := recvMsg.Payload
+
+		// check block header is valid
+		var blockObj types.Block
+		err = rlp.DecodeBytes(block, &blockObj)
+		node.Consensus.PbftLog.AddBlock(&blockObj)
```

Comment:
have you cleaned the blockobj somewhere later?

Reply:
Done.